crates/optimism/rpc/src/eth/transaction.rs

```
@@ -67,6 +73,32 @@ where
 
         Ok(hash)
     }
+
+    async fn send_raw_transaction_sync(
+        &self,
+        tx: Bytes,
+    ) -> Result<Option<RpcReceipt<Self::NetworkTypes>>, Self::Error> {
+        let hash = self.send_raw_transaction(tx).await?;
+        let mut stream = self.provider().canonical_state_stream();
+        match tokio::time::timeout(tokio::time::Duration::from_secs(30), async {
+            while let Some(notification) = stream.next().await {
+                let chain = notification.committed();
+                for block in chain.blocks_iter() {
+                    if block.body().contains_transaction(&hash) {
+                        if let Some(receipt) = self.transaction_receipt(hash).await? {
+                            return Ok(Some(receipt));
+                        }
+                    }
+                }
+            }
+            Ok(None) // Stream ended, no transaction found
+        })
+        .await
+        {
+            Ok(result) => result,
+            Err(_elapsed) => Ok(None), // Timeout reached, no transaction found
```

Small request, but can we return an error with `hash` so that clients can fall back to polling?

We're doing this:
```rust
Err(ErrorObject::owned(
    -32002,
    format!(
        "The transaction was added to the mempool but wasn't processed in {TIMEOUT_DURATION:?}. Please use eth_getTransactionReceipt to poll for the receipt."
    ),
    Some(hash),
))
```

No preferences on the error number and message until this is standardised, so can match if you have better suggestions :pray:.

---

yep make sense